lib/rubocop/sorbet_from_contract_service.rb

```diff
@@ -0,0 +1,169 @@
+# frozen_string_literal: true
+
+module Sorbet
+  class SorbetFromContractService
+    class AutoCorrectError < StandardError; end
+    CONTRACT_PATTERN = RuboCop::NodePattern.new("$(send _ :Contract $... (:hash (:pair $_ $_)))")
+
+    EXTEND_T_SIG_PATTERN = RuboCop::NodePattern.new("(send _ :extend (const (const _ :T) :Sig))")
+
+    CONST_PATTERN = RuboCop::NodePattern.new("(const {nil? (cbase)} $_)")
+    TWO_CONST_PATTERN = RuboCop::NodePattern.new("(const (const {nil? (cbase)} $_) $_)")
+    THREE_CONST_PATTERN = RuboCop::NodePattern.new("(const (const (const {nil? (cbase)} $_) $_) $_)")
+    FOUR_CONST_PATTERN = RuboCop::NodePattern.new("(const (const (const (const {nil? (cbase)} $_) $_) $_) $_)")
+    SEND_PATTERN = RuboCop::NodePattern.new("(send $_ _ $...)")
+    NIL_PATTERN = RuboCop::NodePattern.new("nil")
+    SELF_PATTERN = RuboCop::NodePattern.new("self")
+    HASH_PATTERN = RuboCop::NodePattern.new("(hash $...)")
+    PAIR_MATCHER = RuboCop::NodePattern.new("(pair ({sym str} $_) $_)")
+    CONST_PAIR_MATCHER = RuboCop::NodePattern.new("(hash (pair (const {nil? (cbase)} $_) (const {nil? (cbase)} $_)))")
+    CONTRACT_ARGS_PATTERN = RuboCop::NodePattern.new("({arg optarg blockarg kwoptarg kwarg} $_ ...)")
+    TUPLE_PATTERN = RuboCop::NodePattern.new("(array $...)")
+
+    ARG_NAMES_MATCHER = RuboCop::NodePattern.new("(defs _ _ (:args $...) ...)")
+    INSTANCE_ARG_MATCHER = RuboCop::NodePattern.new("(def _ (:args $...) ...)")
+    PRIVATE_CLASS_ARG_MATCHER = RuboCop::NodePattern.new("(send _ :private_class_method (defs _ _ (:args $...) ...))")
+
+    def self.source(node, args, ret)
+      arg_names = arg_names(node)
+      arg_types = args.map { |arg| convert(arg) }.flatten
+      return_types = convert(ret)
+      format_source(arg_types, arg_names, return_types)
+    rescue AutoCorrectError => e
+      puts e.message
+      nil
+    end
+
+    def self.format_source(arg_types, arg_names, return_types)
```

Comment:
Early returns can make this more elegant:

```ruby
      if arg_names.empty?
        return format("sig { void }", []) if return_types.nil?
        return format("sig { returns(%s) }", return_types)
      end
      
      params = arg_names.zip(arg_types).map do |arg, arg_type|
        arg_name = CONTRACT_ARGS_PATTERN.match(arg).to_s
        "#{arg_name}: #{arg_type}"
      end.join(", ")

      return format("sig { params(%s).void }", params) if return_types.nil?
      format("sig { params(%s).returns(%s) }", params, return_types)
```

Reply:
https://github.com/Shopify/rubocop-sorbet/pull/27/files#diff-f6aa818ee249759bf642b3022cff5d25R34